src/sonic-build-hooks/hooks/git

```diff
@@ -0,0 +1,93 @@
+#!/bin/bash
+
+parse_config(){
+    . /usr/local/share/buildinfo/scripts/buildinfo_base.sh
+    [ -z $REAL_COMMAND ] && REAL_COMMAND=$(get_command git)
+
+    version_file=$VERSION_PATH/versions-git
+    new_version_file=$BUILD_VERSION_PATH/versions-git
+
+    NEEDPROXY=0
+    MODECOMMIT=0
+    whitespace="[[:space:]]"
+    final_str=()
+
+    # parse input parameters
+    for i in "$@"
+    do
+        if [[ $MODECOMMIT == 1 ]];then
+            if [[ $i =~ $whitespace ]];then
+                i=`echo $i | sed 's/ /_/g'`
+            fi
+        fi
+
+        if [[ $i == "clone" ]];then
+            NEEDPROXY=1
+        fi
+
+        # 'git comment -m "comment with space"'. Commands like this will throw exception.
+        if [[ $i == "commit" ]] || [[ $i == "update-ref" ]];then
+            MODECOMMIT=1
+        fi
+
+        final_str+=( "$i" )
+    done
+}
+
+get_clone_path(){
+    # get paremater of clone path
+    while (( "$#" )); do
+        case $1 in
+            -b|--branch|--reference|--reference-if-able|-c|--config|--origin|-u|--upload-pack|-j|--jobs|--depth|--dissociate)
+                shift 2
+                ;;
+            clone|-l|--local|--no-hardlinks|-s|--shared|--dissociate|-q|--quiet|-v|--verbose|--progress|--server-option=*|--bare|--sparse|--filter=*|--template=*|--mirror|--reference|--shallow-*|--no-tags|--recurse-submodules*|--remote-submodules|--no-remote-submodules|--separate-git-dir*)
+                shift 1
+                ;;
+            *)
+                if [[ $URL == "" ]];then
+                    URL=$1
+                else
+                    clone_PATH=$1
+                fi
+                shift 1
+                ;;
+        esac
+    done
+
+    # if not specific clone path, get default clone path
+    [ -z $clone_PATH ] && clone_PATH=`echo $URL | rev | awk -F/ '{print$1}' | rev | awk -F. '{print$1}'`
+}
+
+main(){
+    parse_config $@
+    get_clone_path $@
+
+    # execute git.
+    $REAL_COMMAND ${final_str[*]}
+    result=$?
+
+    # if sub command is not "clone", exit
+    if [[ $NEEDPROXY != 1 ]];then
+        exit $result
+    fi
+
+    pushd $clone_PATH &> /dev/null
+    commit_latest=`$REAL_COMMAND log -n 1 | head -n 1| awk '{print$2}'`
+    [ -f $version_file ] && commit=`grep $URL $version_file | awk -F, '{print$2}'`
+
+    # control version or record version file
+    if [[ $ENABLE_VERSION_CONTROL_GIT == "y" ]];then
+        # control version
+        [ ! -z $commit ] && $REAL_COMMAND reset --hard $commit &>> /dev/null
+    else
+        # record version file
+        echo "$URL==$commit_latest" >> $new_version_file
+        sort $new_version_file -o $new_version_file -u &>> /dev/null
+    fi
+    popd &> /dev/null
+
+    exit $result
+}
+
+main $@
```

Comment:
Change to "$@" ? Do you test the command parameter contains space characters?

Reply:
changed to "$@", and test in bash